kernel/task/src/lib.rs

```diff
@@ -1181,88 +1159,152 @@ impl Drop for JoinableTaskRef {
+    /// If you want to kill another task, use [`TaskRef::kill()`] instead.
     /// 
-    /// This should only be used within task cleanup functions (e.g., after unwinding) to indicate
+    /// This is only usable within task cleanup functions (e.g., after unwinding) to indicate
     /// that the current task has crashed or failed and has been killed by the system.
     /// 
-    /// # Locking / Deadlock
-    /// This method obtains a writable lock on the underlying Task's inner state.
-    /// 
     /// # Return
     /// * Returns `Ok` if the exit status was successfully set.     
     /// * Returns `Err` if this `Task` was already exited, and does not overwrite the existing exit status. 
     ///  
     /// # Note 
     /// The `Task` will not be halted immediately -- 
     /// it will finish running its current timeslice, and then never be run again.
-    #[doc(hidden)]
     pub fn mark_as_killed(&self, reason: KillReason) -> Result<(), &'static str> {
-        if with_current_task(|t| t == self)
-            .unwrap_or(false)
-        {
-            self.internal_exit(ExitValue::Killed(reason))
-        } else {
-            Err("`mark_as_exited()` can only be invoked on the current task, not on another task.")
+        self.internal_exit(ExitValue::Killed(reason))
+    }
+
+    /// Reaps this task (if orphaned) by taking and dropping its exit value and removing it
+    /// from the system task list.
+    ///
+    /// If this task has *not* been orphaned, meaning it is still joinable,
+    /// then this function does nothing.
+    pub fn reap_if_orphaned(&self) {
+        if !self.is_joinable() {
+            // trace!("Reaping orphaned task... {:?}", self);
+            let _exit_value = self.task.reap_exit_value();
+            // trace!("Reaped orphaned task {:?}, {:?}", self, _exit_value);
         }
     }
 
+    /// Perform any actions needed after a context switch.
+    /// 
+    /// Currently this only does two things:
+    /// 1. Drops any data that the original previous task (before the context switch)
+    ///    prepared for us to drop, as specified by `TaskInner::drop_after_task_switch`.
+    /// 2. Obtains the preemption guard such that preemption can be re-enabled
+    ///    when it is appropriate to do so.
+    ///
+    /// Note: this publicly re-exports the private `TaskRef::post_context_switch_action()`
+    ///       function for use in the early `spawn::task_wrapper` functions,
+    ///       which is the only place where an `ExitableTaskRef` can be obtained. 
+    pub fn post_context_switch_action(&self) -> PreemptionGuard {
+        self.task.post_context_switch_action()
+    }
+
+    /// Allows the unwinder to obtain an `ExitableTaskRef` in order for it to
+    /// be able to invoke this task's [`FailureCleanupFunction`].
+    #[doc(hidden)]
+    pub fn obtain_for_unwinder(current_task: TaskRef) -> Self {
+        Self { task: current_task }
+    }
```

Comment:
This should use `get_my_current_task` (or compare `current_task` with `get_my_current_task`) to ensure that we are constructing an `ExitableTaskRef` for the current task.

Reply:
Agreed, good call.